File: Toggly.FeatureManagement.ReactNative/libs/core/tests/TogglyService.test.ts
import { TogglyService } from '../src/services/TogglyService';
import { MemoryStorage } from '../src/services/MemoryStorage';
import type { TogglyConfig, FeatureFlags } from '../src/models';

describe('TogglyService', () => {
  let service: TogglyService;
  let mockStorage: MemoryStorage;

  beforeEach(() => {
    mockStorage = new MemoryStorage();
    jest.clearAllMocks();
  });

  afterEach(() => {
    service?.dispose();
  });

  describe('initialization', () => {
    it('should create instance with default config', () => {
      service = new TogglyService();
      expect(service).toBeInstanceOf(TogglyService);
      expect(service.initialized).toBe(false);
    });

    it('should create instance with custom config', () => {
      const config: TogglyConfig = {
        appKey: 'test-app-key',
        environment: 'Development',
        storage: mockStorage,
      };
      service = new TogglyService(config);
      expect(service).toBeInstanceOf(TogglyService);
    });

    it('should initialize with feature defaults when no appKey provided', async () => {
      const featureDefaults: FeatureFlags = {
        feature1: true,
        feature2: false,
      };

      service = new TogglyService({
        featureDefaults,
        storage: mockStorage,
      });

      const response = await service.init();

      expect(response.status).toBe('defaults');
      expect(response.flags).toEqual(featureDefaults);
      expect(service.initialized).toBe(true);
    });

    it('should generate device ID when no identity provided', async () => {
      service = new TogglyService({
        storage: mockStorage,
      });

      await service.init();

      expect(service.currentIdentity).toBeTruthy();
      expect(typeof service.currentIdentity).toBe('string');
    });

    it('should use provided identity', async () => {
      const identity = 'user-123';
      service = new TogglyService({
        identity,
        storage: mockStorage,
      });

      await service.init();

      expect(service.currentIdentity).toBe(identity);
    });

    it('should fetch features from server when appKey provided', async () => {
      const mockFlags: FeatureFlags = {
        serverFeature: true,
      };

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => mockFlags,
        headers: new Map([['etag', '"abc123"']]),
      });

      service = new TogglyService({
        appKey: 'test-key',
        environment: 'Production',
        storage: mockStorage,
      });

      const response = await service.init();

      expect(response.status).toBe('fetched');
      expect(response.flags).toEqual(mockFlags);
      expect(global.fetch).toHaveBeenCalledTimes(1);
    });

    it('should fall back to defaults on network error', async () => {
      const featureDefaults: FeatureFlags = {
        fallbackFeature: true,
      };

      (global.fetch as jest.Mock).mockRejectedValueOnce(new Error('Network error'));

      service = new TogglyService({
        appKey: 'test-key',
        featureDefaults,
        storage: mockStorage,
      });

      const response = await service.init();

      expect(response.status).toBe('defaults');
      expect(response.flags).toEqual(featureDefaults);
      expect(response.error).toBe('Network error');
    });
  });

  describe('feature evaluation', () => {
    beforeEach(async () => {
      service = new TogglyService({
        featureDefaults: {
          enabledFeature: true,
          disabledFeature: false,
          feature1: true,
          feature2: true,
          feature3: false,
        },
        storage: mockStorage,
      });
      await service.init();
    });

    describe('isFeatureOn', () => {
      it('should return true for enabled feature', async () => {
        const result = await service.isFeatureOn('enabledFeature');
        expect(result).toBe(true);
      });

      it('should return false for disabled feature', async () => {
        const result = await service.isFeatureOn('disabledFeature');
        expect(result).toBe(false);
      });

      it('should return false for unknown feature', async () => {
        const result = await service.isFeatureOn('unknownFeature');
        expect(result).toBe(false);
      });
    });

    describe('isFeatureOff', () => {
      it('should return false for enabled feature', async () => {
        const result = await service.isFeatureOff('enabledFeature');
        expect(result).toBe(false);
      });

      it('should return true for disabled feature', async () => {
        const result = await service.isFeatureOff('disabledFeature');
        expect(result).toBe(true);
      });

      it('should return true for unknown feature', async () => {
        const result = await service.isFeatureOff('unknownFeature');
        expect(result).toBe(true);
      });
    });

    describe('evaluateFeatureGate', () => {
      it('should return true when all features are enabled (all requirement)', async () => {
        const result = await service.evaluateFeatureGate(
          ['feature1', 'feature2'],
          'all'
        );
        expect(result).toBe(true);
      });

      it('should return false when some features are disabled (all requirement)', async () => {
        const result = await service.evaluateFeatureGate(
          ['feature1', 'feature3'],
          'all'
        );
        expect(result).toBe(false);
      });

      it('should return true when any feature is enabled (any requirement)', async () => {
        const result = await service.evaluateFeatureGate(
          ['feature1', 'feature3'],
          'any'
        );
        expect(result).toBe(true);
      });

      it('should return false when no features are enabled (any requirement)', async () => {
        const result = await service.evaluateFeatureGate(
          ['feature3', 'unknownFeature'],
          'any'
        );
        expect(result).toBe(false);
      });

      it('should negate the result when negate is true', async () => {
        const result = await service.evaluateFeatureGate(
          ['enabledFeature'],
          'all',
          true
        );
        expect(result).toBe(false);
      });

      it('should return true for empty feature keys', async () => {
        const result = await service.evaluateFeatureGate([]);
        expect(result).toBe(true);
      });
    });
  });

  describe('identity management', () => {
    it('should update identity and refresh features', async () => {
      service = new TogglyService({
        featureDefaults: { feature1: true },
        storage: mockStorage,
      });
      await service.init();

      const initialIdentity = service.currentIdentity;
      const newIdentity = 'new-user-456';

      const response = await service.setIdentity(newIdentity);

      expect(service.currentIdentity).toBe(newIdentity);
      expect(service.currentIdentity).not.toBe(initialIdentity);
      expect(response.status).toBe('defaults');
    });

    it('should revert to device ID when identity is null', async () => {
      service = new TogglyService({
        identity: 'user-123',
        storage: mockStorage,
      });
      await service.init();

      await service.setIdentity(null);

      expect(service.currentIdentity).toBeTruthy();
      expect(service.currentIdentity).not.toBe('user-123');
    });
  });

  describe('caching', () => {
    it('should cache features after fetch', async () => {
      const mockFlags: FeatureFlags = { cachedFeature: true };

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => mockFlags,
        headers: new Map(),
      });

      service = new TogglyService({
        appKey: 'test-key',
        storage: mockStorage,
      });

      await service.init();

      // Verify cache was written
      const keys = mockStorage.keys();
      expect(keys.some((k) => k.startsWith('@toggly:featureFlagsCache:'))).toBe(true);
    });

    it('should clear cache when requested', async () => {
      service = new TogglyService({
        featureDefaults: { feature1: true },
        storage: mockStorage,
      });
      await service.init();

      await service.clearCache();

      expect(service.currentFeatures).toBeNull();
    });
  });

  describe('hooks', () => {
    it('should register and execute hooks', async () => {
      const beforeEvaluationMock = jest.fn();
      const afterEvaluationMock = jest.fn();

      service = new TogglyService({
        featureDefaults: { testFeature: true },
        hooks: [
          {
            getMetadata: () => ({ name: 'TestHook' }),
            beforeEvaluation: beforeEvaluationMock,
            afterEvaluation: afterEvaluationMock,
          },
        ],
        storage: mockStorage,
      });

      await service.init();
      await service.isFeatureOn('testFeature');

      expect(beforeEvaluationMock).toHaveBeenCalledWith('testFeature', undefined);
      expect(afterEvaluationMock).toHaveBeenCalledWith('testFeature', undefined, true);
    });

    it('should add hook dynamically', async () => {
      const hookMock = jest.fn();

      service = new TogglyService({
        featureDefaults: { testFeature: true },
        storage: mockStorage,
      });
      await service.init();

      service.addHook({
        getMetadata: () => ({ name: 'DynamicHook' }),
        afterEvaluation: hookMock,
      });

      await service.isFeatureOn('testFeature');

      expect(hookMock).toHaveBeenCalled();
    });

    it('should remove hook by name', async () => {
      const hookMock = jest.fn();

      service = new TogglyService({
        featureDefaults: { testFeature: true },
        hooks: [
          {
            getMetadata: () => ({ name: 'RemovableHook' }),
            afterEvaluation: hookMock,
          },
        ],
        storage: mockStorage,
      });
      await service.init();

      const removed = service.removeHook('RemovableHook');
      expect(removed).toBe(true);

      await service.isFeatureOn('testFeature');
      expect(hookMock).not.toHaveBeenCalled();
    });
  });

  describe('events', () => {
    it('should emit initialized event', async () => {
      const listener = jest.fn();

      service = new TogglyService({
        featureDefaults: { feature1: true },
        storage: mockStorage,
      });

      service.on('initialized', listener);
      await service.init();

      expect(listener).toHaveBeenCalledTimes(1);
      expect(listener).toHaveBeenCalledWith(
        expect.objectContaining({
          type: 'initialized',
          timestamp: expect.any(Date),
        })
      );
    });

    it('should emit identityChanged event', async () => {
      const listener = jest.fn();

      service = new TogglyService({
        identity: 'user-1',
        storage: mockStorage,
      });
      await service.init();

      service.on('identityChanged', listener);
      await service.setIdentity('user-2');

      expect(listener).toHaveBeenCalledWith(
        expect.objectContaining({
          type: 'identityChanged',
          data: {
            previousIdentity: 'user-1',
            newIdentity: 'user-2',
          },
        })
      );
    });

    it('should unsubscribe from events', async () => {
      const listener = jest.fn();

      service = new TogglyService({
        storage: mockStorage,
      });

      const unsubscribe = service.on('initialized', listener);
      unsubscribe();

      await service.init();

      expect(listener).not.toHaveBeenCalled();
    });
  });

  describe('state change handlers', () => {
    it('should notify handlers when features change', async () => {
      const handler = jest.fn();

      (global.fetch as jest.Mock)
        .mockResolvedValueOnce({
          ok: true,
          status: 200,
          json: async () => ({ feature1: true }),
          headers: new Map(),
        })
        .mockResolvedValueOnce({
          ok: true,
          status: 200,
          json: async () => ({ feature1: false }),
          headers: new Map(),
        });

      service = new TogglyService({
        appKey: 'test-key',
        storage: mockStorage,
      });

      await service.init();
      service.addStateChangeHandler(handler);
      await service.refresh();

      expect(handler).toHaveBeenCalledWith('feature1', true, false);
    });

    it('should remove state change handler', async () => {
      const handler = jest.fn();

      service = new TogglyService({
        featureDefaults: { feature1: true },
        storage: mockStorage,
      });
      await service.init();

      const removeHandler = service.addStateChangeHandler(handler);
      removeHandler();

      // Handler should not be called after removal
      expect(handler).not.toHaveBeenCalled();
    });
  });

  describe('debug info', () => {
    it('should return debug information', async () => {
      service = new TogglyService({
        appKey: 'test-key',
        environment: 'Production',
        identity: 'user-123',
        storage: mockStorage,
      });
      await service.init();

      const debugInfo = service.getDebugInfo();

      expect(debugInfo.appKey).toBe('test-key');
      expect(debugInfo.environment).toBe('Production');
      expect(debugInfo.identity).toBe('user-123');
      expect(debugInfo.syncServiceRunning).toBe(true);
    });
  });

  describe('dispose', () => {
    it('should clean up resources on dispose', async () => {
      const listener = jest.fn();

      service = new TogglyService({
        featureDefaults: { feature1: true },
        storage: mockStorage,
      });
      await service.init();

      service.on('refreshed', listener);
      service.dispose();

      expect(service.initialized).toBe(false);
      expect(service.currentFeatures).toBeNull();
    });
  });

  describe('network and app state', () => {
    it('should handle network state changes', async () => {
      let networkCallback: ((state: { isConnected: boolean }) => void) | null = null;
      const mockNetworkInfo = {
        subscribe: jest.fn((callback) => {
          networkCallback = callback;
          return () => {};
        }),
        getState: jest.fn().mockResolvedValue({ isConnected: true }),
      };

      service = new TogglyService({
        featureDefaults: { feature1: true },
        storage: mockStorage,
        networkInfo: mockNetworkInfo,
      });

      await service.init();
      expect(mockNetworkInfo.subscribe).toHaveBeenCalled();

      // Simulate going offline then online
      if (networkCallback) {
        networkCallback({ isConnected: false });
        networkCallback({ isConnected: true });
      }
    });

    it('should handle app state changes', async () => {
      let appStateCallback: ((state: string) => void) | null = null;
      const mockAppState = {
        getCurrentState: jest.fn().mockReturnValue('active'),
        subscribe: jest.fn((callback) => {
          appStateCallback = callback;
          return () => {};
        }),
      };

      service = new TogglyService({
        featureDefaults: { feature1: true },
        storage: mockStorage,
        appState: mockAppState,
      });

      await service.init();
      expect(mockAppState.subscribe).toHaveBeenCalled();

      // Simulate going to background then foreground
      if (appStateCallback) {
        appStateCallback('background');
        appStateCallback('active');
      }
    });

    it('should skip refresh when app is in background', async () => {
      const mockAppState = {
        getCurrentState: jest.fn().mockReturnValue('background'),
        subscribe: jest.fn(() => () => {}),
      };

      service = new TogglyService({
        appKey: 'test-key',
        featureDefaults: { feature1: true },
        storage: mockStorage,
        appState: mockAppState,
      });

      await service.init();

      // Refresh should return cached because app is in background
      const response = await service.refresh();
      expect(response.status).toBe('cached');
    });

    it('should skip refresh when offline', async () => {
      const mockNetworkInfo = {
        subscribe: jest.fn(() => () => {}),
        getState: jest.fn().mockResolvedValue({ isConnected: false }),
      };

      service = new TogglyService({
        appKey: 'test-key',
        featureDefaults: { feature1: true },
        storage: mockStorage,
        networkInfo: mockNetworkInfo,
      });

      await service.init();

      // Manually set network state to offline
      (service as any).networkState = { isConnected: false };

      const response = await service.refresh();
      expect(response.status).toBe('cached');
    });
  });

  describe('API responses', () => {
    it('should handle 304 Not Modified response', async () => {
      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => ({ feature1: true }),
        headers: new Map([['etag', '"abc123"']]),
      }).mockResolvedValueOnce({
        ok: false,
        status: 304,
        statusText: 'Not Modified',
      });

      service = new TogglyService({
        appKey: 'test-key',
        storage: mockStorage,
      });

      await service.init();

      // Second refresh should handle 304
      const response = await service.refresh();
      expect(response.status).toBe('cached');
    });

    it('should handle HTTP errors', async () => {
      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: false,
        status: 500,
        statusText: 'Internal Server Error',
      });

      service = new TogglyService({
        appKey: 'test-key',
        featureDefaults: { fallback: true },
        storage: mockStorage,
      });

      const response = await service.init();
      expect(response.status).toBe('defaults');
      expect(response.error).toContain('500');
    });

    it('should handle signed definitions response', async () => {
      const mockSignedData = {
        data: { signedFeature: true },
        signature: 'mock-signature',
      };

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => mockSignedData,
        headers: new Map(),
      });

      service = new TogglyService({
        appKey: 'test-key',
        useSignedDefinitions: true,
        storage: mockStorage,
      });

      const response = await service.init();
      expect(response.status).toBe('fetched');
      expect(response.flags).toEqual({ signedFeature: true });
    });

    it('should add If-None-Match header when ETag exists', async () => {
      (global.fetch as jest.Mock)
        .mockResolvedValueOnce({
          ok: true,
          status: 200,
          json: async () => ({ feature1: true }),
          headers: new Map([['etag', '"abc123"']]),
        })
        .mockResolvedValueOnce({
          ok: true,
          status: 200,
          json: async () => ({ feature1: true }),
          headers: new Map(),
        });

      service = new TogglyService({
        appKey: 'test-key',
        useSignedDefinitions: true,
        storage: mockStorage,
      });

      await service.init();
      await service.refresh();

      // Check that fetch was called with the If-None-Match header
      expect(global.fetch).toHaveBeenCalledTimes(2);
    });
  });

  describe('concurrent requests', () => {
    it('should prevent duplicate fetches', async () => {
      let resolvePromise: () => void;
      const slowPromise = new Promise<void>((resolve) => {
        resolvePromise = resolve;
      });

      (global.fetch as jest.Mock).mockImplementation(() =>
        slowPromise.then(() => ({
          ok: true,
          status: 200,
          json: async () => ({ feature1: true }),
          headers: new Map(),
        }))
      );

      service = new TogglyService({
        appKey: 'test-key',
        storage: mockStorage,
      });

      // Start two concurrent refreshes
      const promise1 = service.init();
      const promise2 = service.refresh();

      // Resolve the fetch
      resolvePromise!();

      await Promise.all([promise1, promise2]);

      // Should only have made one fetch call
      expect(global.fetch).toHaveBeenCalledTimes(1);
    });
  });

  describe('onAll event listener', () => {
    it('should receive all events', async () => {
      const allEventsListener = jest.fn();

      service = new TogglyService({
        featureDefaults: { feature1: true },
        storage: mockStorage,
      });

      service.onAll(allEventsListener);
      await service.init();
      await service.setIdentity('user-1');

      expect(allEventsListener).toHaveBeenCalled();
    });
  });

  describe('feature evaluation edge cases', () => {
    it('should handle single feature in gate with negate', async () => {
      service = new TogglyService({
        featureDefaults: { feature1: true },
        storage: mockStorage,
      });
      await service.init();

      const result = await service.evaluateFeatureGate(['feature1'], 'all', true);
      expect(result).toBe(false);
    });

    it('should use defaults when features is null', async () => {
      service = new TogglyService({
        featureDefaults: { defaultFeature: true },
        storage: mockStorage,
      });

      // Don't init, evaluate directly
      const result = await service.isFeatureOn('defaultFeature');
      expect(result).toBe(true);
    });
  });

  describe('showFeatureDuringEvaluation', () => {
    it('should return the configured value', () => {
      service = new TogglyService({
        showFeatureDuringEvaluation: true,
        storage: mockStorage,
      });

      expect(service.shouldShowFeatureDuringEvaluation).toBe(true);
    });
  });

  describe('state change handler errors', () => {
    it('should continue when handler throws error', async () => {
      const errorHandler = jest.fn().mockImplementation(() => {
        throw new Error('Handler error');
      });
      const goodHandler = jest.fn();
      const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();

      (global.fetch as jest.Mock)
        .mockResolvedValueOnce({
          ok: true,
          status: 200,
          json: async () => ({ feature1: true }),
          headers: new Map(),
        })
        .mockResolvedValueOnce({
          ok: true,
          status: 200,
          json: async () => ({ feature1: false }),
          headers: new Map(),
        });

      service = new TogglyService({
        appKey: 'test-key',
        storage: mockStorage,
      });

      await service.init();
      service.addStateChangeHandler(errorHandler);
      service.addStateChangeHandler(goodHandler);
      await service.refresh();

      expect(errorHandler).toHaveBeenCalled();
      expect(goodHandler).toHaveBeenCalled();
      expect(consoleErrorSpy).toHaveBeenCalled();

      consoleErrorSpy.mockRestore();
    });
  });
});

File: Toggly.FeatureManagement.Android/toggly-android-core/src/test/kotlin/io/toggly/core/TogglyServiceTest.kt
package io.toggly.core

import io.toggly.core.models.*
import io.toggly.core.storage.MemoryStorage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.test.*
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

@OptIn(ExperimentalCoroutinesApi::class)
class TogglyServiceTest {

    private lateinit var service: TogglyService
    private lateinit var storage: MemoryStorage
    private val testDispatcher = StandardTestDispatcher()

    @Before
    fun setup() {
        Dispatchers.setMain(testDispatcher)
        storage = MemoryStorage()
    }

    @After
    fun tearDown() {
        Dispatchers.resetMain()
    }

    private fun createService(
        appKey: String? = "test-app-key",
        environment: String = "Production",
        featureDefaults: FeatureFlags = emptyMap(),
        refreshInterval: Long = 0
    ): TogglyService {
        val config = TogglyConfig(
            appKey = appKey,
            environment = environment,
            featureDefaults = featureDefaults,
            refreshInterval = refreshInterval,
            storage = storage
        )
        return TogglyService(config)
    }

    @Test
    fun `service initializes with default flags`() = runTest {
        val defaults = mapOf("feature1" to true, "feature2" to false)
        service = createService(featureDefaults = defaults)

        assertTrue(service.isFeatureOn("feature1"))
        assertFalse(service.isFeatureOn("feature2"))
    }

    @Test
    fun `isFeatureOn returns true for enabled feature`() = runTest {
        val defaults = mapOf("enabled-feature" to true)
        service = createService(featureDefaults = defaults)

        assertTrue(service.isFeatureOn("enabled-feature"))
    }

    @Test
    fun `isFeatureOn returns false for disabled feature`() = runTest {
        val defaults = mapOf("disabled-feature" to false)
        service = createService(featureDefaults = defaults)

        assertFalse(service.isFeatureOn("disabled-feature"))
    }

    @Test
    fun `isFeatureOn returns false for unknown feature`() = runTest {
        service = createService()

        assertFalse(service.isFeatureOn("unknown-feature"))
    }

    @Test
    fun `isFeatureOff returns opposite of isFeatureOn`() = runTest {
        val defaults = mapOf("enabled" to true, "disabled" to false)
        service = createService(featureDefaults = defaults)

        assertTrue(service.isFeatureOff("disabled"))
        assertFalse(service.isFeatureOff("enabled"))
        assertTrue(service.isFeatureOff("unknown"))
    }

    @Test
    fun `evaluateFeatureGate with ALL requirement - all enabled`() = runTest {
        val defaults = mapOf(
            "feature1" to true,
            "feature2" to true,
            "feature3" to true
        )
        service = createService(featureDefaults = defaults)

        val result = service.evaluateFeatureGate(
            listOf("feature1", "feature2", "feature3"),
            FeatureRequirement.ALL
        )

        assertTrue(result)
    }

    @Test
    fun `evaluateFeatureGate with ALL requirement - one disabled`() = runTest {
        val defaults = mapOf(
            "feature1" to true,
            "feature2" to false,
            "feature3" to true
        )
        service = createService(featureDefaults = defaults)

        val result = service.evaluateFeatureGate(
            listOf("feature1", "feature2", "feature3"),
            FeatureRequirement.ALL
        )

        assertFalse(result)
    }

    @Test
    fun `evaluateFeatureGate with ANY requirement - one enabled`() = runTest {
        val defaults = mapOf(
            "feature1" to false,
            "feature2" to true,
            "feature3" to false
        )
        service = createService(featureDefaults = defaults)

        val result = service.evaluateFeatureGate(
            listOf("feature1", "feature2", "feature3"),
            FeatureRequirement.ANY
        )

        assertTrue(result)
    }

    @Test
    fun `evaluateFeatureGate with ANY requirement - all disabled`() = runTest {
        val defaults = mapOf(
            "feature1" to false,
            "feature2" to false,
            "feature3" to false
        )
        service = createService(featureDefaults = defaults)

        val result = service.evaluateFeatureGate(
            listOf("feature1", "feature2", "feature3"),
            FeatureRequirement.ANY
        )

        assertFalse(result)
    }

    @Test
    fun `evaluateFeatureGate with negate - inverts result`() = runTest {
        val defaults = mapOf("feature1" to true)
        service = createService(featureDefaults = defaults)

        val result = service.evaluateFeatureGate(
            listOf("feature1"),
            FeatureRequirement.ALL,
            negate = true
        )

        assertFalse(result)
    }

    @Test
    fun `evaluateFeatureGate with empty list returns true for ALL`() = runTest {
        service = createService()

        val result = service.evaluateFeatureGate(
            emptyList(),
            FeatureRequirement.ALL
        )

        assertTrue(result)
    }

    @Test
    fun `evaluateFeatureGate with empty list returns true for ANY`() = runTest {
        service = createService()

        val result = service.evaluateFeatureGate(
            emptyList(),
            FeatureRequirement.ANY
        )

        // Empty list returns true unconditionally
        assertTrue(result)
    }

    @Test
    fun `featureFlagFlow emits current value`() = runTest {
        val defaults = mapOf("feature1" to true)
        service = createService(featureDefaults = defaults)

        val result = service.featureFlagFlow("feature1").first()
        assertTrue(result)
    }

    @Test
    fun `featureFlagFlow emits false for unknown feature`() = runTest {
        service = createService()

        val result = service.featureFlagFlow("unknown").first()
        assertFalse(result)
    }

    @Test
    fun `setIdentity updates current identity`() = runTest {
        service = createService()
        service.setIdentity("user-123")

        assertEquals("user-123", service.currentIdentity)
    }

    @Test
    fun `setIdentity with null uses device ID`() = runTest {
        service = createService()
        service.setIdentity("user-123")
        service.setIdentity(null)

        // When null is passed, service uses or generates device ID
        assertNotNull(service.currentIdentity)
        // Identity should not be the previously set value
        assertNotEquals("user-123", service.currentIdentity)
    }

    @Test
    fun `clearCache removes Toggly cache keys`() = runTest {
        service = createService()
        // Set a Toggly-specific cache key
        storage.set("@toggly:etag", "test-etag-value")

        service.clearCache()

        // Toggly cache keys should be cleared
        assertNull(storage.get("@toggly:etag"))
    }

    @Test
    fun `getDebugInfo returns current state`() = runTest {
        val defaults = mapOf("feature1" to true)
        service = createService(
            appKey = "test-key",
            environment = "staging",
            featureDefaults = defaults
        )
        service.setIdentity("user-123")

        val debugInfo = service.getDebugInfo()

        assertEquals("test-key", debugInfo.appKey)
        assertEquals("staging", debugInfo.environment)
        assertEquals("user-123", debugInfo.identity)
    }

    @Test
    fun `events flow is available`() = runTest {
        service = createService()

        // Just verify the flow exists and can be collected
        assertNotNull(service.events)
    }

    @Test
    fun `service with environment includes it in debug info`() = runTest {
        service = createService(environment = "staging")

        val debugInfo = service.getDebugInfo()
        assertEquals("staging", debugInfo.environment)
    }

    @Test
    fun `feature gate flow emits correct values`() = runTest {
        val defaults = mapOf(
            "feature1" to true,
            "feature2" to true
        )
        service = createService(featureDefaults = defaults)

        val result = service.featureGateFlow(
            listOf("feature1", "feature2"),
            FeatureRequirement.ALL
        ).first()

        assertTrue(result)
    }

    @Test
    fun `multiple features can be checked independently`() = runTest {
        val defaults = mapOf(
            "a" to true,
            "b" to false,
            "c" to true
        )
        service = createService(featureDefaults = defaults)

        assertTrue(service.isFeatureOn("a"))
        assertFalse(service.isFeatureOn("b"))
        assertTrue(service.isFeatureOn("c"))
    }

    @Test
    fun `default flags are used for evaluation when cache is empty`() = runTest {
        val defaults = mapOf("default-feature" to true)
        service = createService(featureDefaults = defaults)

        // Default flags are used for feature evaluation
        assertTrue(service.isFeatureOn("default-feature"))
    }

    @Test
    fun `service handles special characters in feature keys`() = runTest {
        val defaults = mapOf(
            "feature:with:colons" to true,
            "feature.with.dots" to true,
            "feature-with-dashes" to true,
            "feature_with_underscores" to true
        )
        service = createService(featureDefaults = defaults)

        assertTrue(service.isFeatureOn("feature:with:colons"))
        assertTrue(service.isFeatureOn("feature.with.dots"))
        assertTrue(service.isFeatureOn("feature-with-dashes"))
        assertTrue(service.isFeatureOn("feature_with_underscores"))
    }

    @Test
    fun `case sensitivity is maintained for feature keys`() = runTest {
        val defaults = mapOf(
            "Feature" to true,
            "feature" to false,
            "FEATURE" to true
        )
        service = createService(featureDefaults = defaults)

        assertTrue(service.isFeatureOn("Feature"))
        assertFalse(service.isFeatureOn("feature"))
        assertTrue(service.isFeatureOn("FEATURE"))
    }
}
